### ISSUE TITLE
kubernetes auth not working

### ISSUE THREAD

**Issue by kyuwoo-choi** [ricoberger/vault-secrets-operator#184]
I'm looking for the solution to the k8s auth error.
The logs are below from vault-secrets-operator / vault
From the vault log, I suspect something wrong with the req. The jwt and role values look wired.
shouldn't the value be like `role: vault-secrets-operator`?

I installed vault-secrets-operator from helm, and installed in the same namespace to the vault.

## vault-secrets-operator logs
```
vault-secrets-operator {"level":"info","ts":"2023-03-14T03:14:50Z","logger":"vault","msg":"Reconciliation is enabled.","ReconciliationTime":0}
vault-secrets-operator {"level":"error","ts":"2023-03-14T03:14:50Z","msg":"Could not create API client for Vault","error":"Error making API request.\n\nURL: PUT http://vault:8200/v1/auth/kubernetes/login\nCode: 403. Errors:\n\n* service account name not authorized","stacktrace":"main.main\n\t/workspace/main.go:56\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
```

## vault logs
```
vault {"time":"2023-03-14T03:03:49.891628388Z","type":"request","auth":{"policy_results":{"allowed":true},"token_type":"default"},"request":{"id":"dcbe89c4-4871-82c2-ba8e-f7c2635d50a7","operation":"update","mount_type":"kubernetes","mount_accessor":"auth_kubernetes_562c3377","namespace":{"id":"root"},"path":"auth/kubernetes/login","data":{"jwt":"hmac-sha256:0267709647518c9a8c1b8f6f1de445c356eeaa17a87e2d65cb183dbb4e9900b5","role":"hmac-sha256:48b42f5dc943aca4646a0433a8f8d6ae45330f9f37b58dadbbdedd056c09259c"},"remote_address":"127.0.0.6","remote_port":55383}}
vault {"time":"2023-03-14T03:03:49.892112466Z","type":"response","auth":{"policy_results":{"allowed":true},"token_type":"default"},"request":{"id":"dcbe89c4-4871-82c2-ba8e-f7c2635d50a7","operation":"update","mount_type":"kubernetes","mount_accessor":"auth_kubernetes_562c3377","namespace":{"id":"root"},"path":"auth/kubernetes/login","data":{"jwt":"hmac-sha256:0267709647518c9a8c1b8f6f1de445c356eeaa17a87e2d65cb183dbb4e9900b5","role":"hmac-sha256:48b42f5dc943aca4646a0433a8f8d6ae45330f9f37b58dadbbdedd056c09259c"},"remote_address":"127.0.0.6","remote_port":55383},"response":{"mount_type":"kubernetes","mount_accessor":"auth_kubernetes_562c3377"},"error":"service account name not authorized"}
```

What can I do to fix this?

**Comment by ricoberger** [comment]
Hi @kyuwoo-choi, did you run the commands from the readme to authorize the operator to access Vault: https://github.com/ricoberger/vault-secrets-operator#kubernetes-auth-method

**Comment by nimeshkummar** [comment]
i am on GKE and i followed all the steps. still getting the same error as mentioned above

**Comment by ricoberger** [comment]
Can you share the Kubernetes + Vault version you are using please?

**Comment by nimeshkummar** [comment]
Vault - 1.13.2
K8s - 1.24

**Comment by ricoberger** [comment]
Thanks for the details @michalgoldys, that was super helpful 🙂.

I checked the tests and it seems that the `SA_JWT_TOKEN` is not available anymore (not sure why it's still working in the tests): https://github.com/ricoberger/vault-secrets-operator/actions/runs/4918785370/jobs/8785564485?pr=190

I think generating the secret via the Helm chart should be fine and I created a PR for this #191. At the moment the `serviceAccount.createSecret` value is set to `false` by default to not break existing setups, but I'm also ok with changing this to `true`. What do you think?

With the created secret the `SA_JWT_TOKEN` is also available again in the tests: https://github.com/ricoberger/vault-secrets-operator/actions/runs/4918996956/jobs/8786055265?pr=191

**Comment by michalgoldys** [comment]
No problem, that is something that we were fighting a bit recently. 

Generally, based on:
https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
`LegacyServiceAccountTokenNoAutoGeneration` was `Beta` and `Enabled` by Default from `v1.24`, until `v1.25`, and from `v1.26` it is GA. 

So either since `v1.24` secret generation should be enabled by default (since it no longer would be generated automatically) OR TokenRequest API should be used, as recommended:
https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/

```
Note: Although the manual mechanism for creating a long-lived ServiceAccount token exists, using [TokenRequest](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/) to obtain short-lived API access tokens is recommended instead.
```

There are also some changes in v1.27 according to TokenRequests:
https://kubernetes.io/docs/concepts/configuration/secret/#service-account-token-secrets
```
In more recent versions, including Kubernetes v1.27, API credentials are obtained directly by using the [TokenRequest](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/) API, and are mounted into Pods using a [projected volume](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#bound-service-account-token-volume). The tokens obtained using this method have bounded lifetimes, and are automatically invalidated when the Pod they are mounted into is deleted.
```

But returning to your question, either creating a secret should be enabled by default from v1.24.
In ServiceAccount and Secret approach would need to ensure from which Kubernetes version annotation `type: kubernetes.io/service-account-token` is working, to not break anything back ;)

**Comment by ricoberger** [comment]
Thanks again @michalgoldys 🙂

@nimeshkummar @kyuwoo-choi the issue should be fixed in version 2.4.0 of the Helm chart.
